### ISSUE TITLE
Consolidate Redundant Functions into a Single Library

### ISSUE THREAD

**Issue by KyroChi** [amitsi/pluribus-ansible#196]
I have noticed a lot of redundant code in this code base. For example, the following code is pulled from two **SEPERATE** files, and is identical in every way:

```python
pn_ztp_vrrp_l3.py
...
def create_cluster(module, switch, name, node1, node2):
    ...
    global CHANGED_FLAG
    cli = pn_cli(module)
    clicopy = cli
    cli += ' switch %s cluster-show format name no-show-headers ' % node1
    cluster_list = run_cli(module, cli).split()
    if name not in cluster_list:
        cli = clicopy
        cli += ' switch %s cluster-create name %s ' % (switch, name)
        cli += ' cluster-node-1 %s cluster-node-2 %s ' % (node1, node2)
        if 'Success' in run_cli(module, cli):
            CHANGED_FLAG.append(True)
            return ' %s: %s created successfully \n' % (switch, name)
    else:
        return ' %s: %s already exists \n' % (switch, name)
...
```
```python
pn_ztp_vrrp_l3_additional_switches.py
...
def create_cluster(module, switch, name, node1, node2):
    ...
    global CHANGED_FLAG
    cli = pn_cli(module)
    clicopy = cli
    cli += ' switch %s cluster-show format name no-show-headers ' % node1
    cluster_list = run_cli(module, cli).split()
    if name not in cluster_list:
        cli = clicopy
        cli += ' switch %s cluster-create name %s ' % (switch, name)
        cli += ' cluster-node-1 %s cluster-node-2 %s ' % (node1, node2)
        if 'Success' in run_cli(module, cli):
            CHANGED_FLAG.append(True)
            return ' %s: %s created successfully \n' % (switch, name)
    else:
        return ' %s: %s already exists \n' % (switch, name)
...
```

This appears to be an actual copy and paste between files, and this kind of thing isn't just with one function, it appears in almost every file, with a long list of functions. This is dangerous, because if we find a bug in the cluster_create function, we have to change it in every single location, instead of one single location.

What I propose is that we create a library which we import at the top of our modules that contains commonly used functions such as create_cluster and create_vlan to name a few. This will make it so any bug discovered in the future can be fixed **ONCE**, and it will be fixed throughout the entire codebase. This change will also significantly reduce the size of the codebase, which will make the code more readable, and will contribute (slightly) to faster execution time. In addition, it will streamline debugging time, and make writing new modules much easier, meaning we can push more features at a faster rate.

**Comment by amitsi** [comment]
I agree, this was some of the feedback last time when we pushed code to ansible....can you go through all the modules and come up with a list of functions

**Comment by KyroChi** [comment]
@amitsi I am building a library for all the shared functions. I am going to go through all of the files and refactor them to use the new library for all the functions. I probably won't be done until Tuesday, maybe Friday afternoon. Its going to be a large refactoring, but it should make development of new modules much easier. I will include a write up with a guide to using the refactored code to create new modules. 

I will post updates to this issue if anything comes up.

**Comment by amitsi** [comment]
take till wednesday as we need to do it for all.....let's sync up before you make a PR as we need to figure out if we need to push across branches too @ashxjain let's talk later today or monday

**Comment by ashxjain** [comment]
Thanks @KyroChi. This is much needed. But make sure that we write it in a way acceptable by Ansible Guidelines.

**Comment by KyroChi** [comment]
@ashxjain Do you have a link to their guidelines? I have been unsuccessful in my Google search. Guessing I am searching for the wrong thing or just a fool. :smile: Thanks.

**Comment by amitsi** [comment]
@KyroChi yes...the guidelines are on slack on lab-information channel....let me get it for you

**Comment by KyroChi** [comment]
https://docs.ansible.com/ansible/dev_guide/developing_modules_in_groups.html

**Comment by ashxjain** [comment]
Ok. I think so this where all the common code goes: https://github.com/ansible/ansible/tree/devel/lib/ansible/module_utils

**Comment by KyroChi** [comment]
@ashxjain For the time being I think we will want to make a script to prepend the library file to our files. We should have everything running and up to Ansible standards before we try and push anything to the Ansible repo.

**Comment by ashxjain** [comment]
@KyroChi You are right. But lets follow the way others have done in module_utils directory.

**Comment by KyroChi** [comment]
Here are some files that give a good idea of how the new framework will work:

[Framework](https://github.com/amitsi/pluribus-ansible/blob/kchick/ansible/library/pn_ansible_lib.py)
[Unit Tests](https://github.com/amitsi/pluribus-ansible/blob/kchick/pn_ansible_lib_tests.py)

Let me know if you have any comments, I will make a code review when I have some more.

**Comment by ashxjain** [comment]
Hi @KyroChi,
I took a quick look at the code, it looks great. Thanks for adding those unit tests, it'll be very helpful.
I'll soon give in-detailed review.

@gaurav-calsoft @saptaktakalkar Do review these changes and see if there's anything else to be done here.

Thanks!

**Comment by KyroChi** [comment]
Hello all,

Keep in mind this is sort of a preview. I will have a lot more coming next week. Any preliminary comments will be appreciated though

Thanks
- Kyle

**Comment by KyroChi** [comment]
@amitsi @ashxjain Here is a quick "snapshot" commit. [link](https://github.com/amitsi/pluribus-ansible/commit/e08bad55e061a9fab89b50b1ec7d3b911769f5b3)

Also here is some unfinished documentation. [link](https://github.com/amitsi/pluribus-ansible/wiki/Developing-with-pn_ansible_lib.py)

Will update with more later in the week when this is all done.
